### PR TITLE
✨ generate csv and zip file server side

### PR DESCRIPTION
**Pull request by danyx23**
This PR adds functionality to generate a CSV file on the server in a CF worker for the data of any chart. It also allows downloading a metadata.json file, a readme.md and a zip file of all three of these things. 

This is in preparation of surfacing these things in the download UI of grapher in the upcoming cycle 2024.6 (#4015). This PR does not make any use of the new CF functions endpoints yet and the download UI is not yet changed. 

This PR implements #3648 

## Testing

To test this, try http get requests against `/grapher/SLUG.zip`, `/grapher/SLUG.csv`, `/grapher/SLUG.metadata.json` at localhost:8788, or at the staging server linked below.

There is also an observable notebook that lets you browse the generated readme files in an easier way: https://observablehq.com/d/d410e9b2d2b7c330

### CODE REVIEW

File: functions/grapher/csv/[slug].ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+const router = Router<IRequestStrict, [URL, Env, ExecutionContext]>()
+router
+    .get(
+        "/grapher/csv/:slug",
+        async ({ params: { slug } }, { searchParams: _ }, env) =>
+            fetchCsvForGrapher(slug, env)
+    )
+    .all("*", () => error(404, "Route not defined"))
```
Comment (larsyencken) [inline functions/grapher/csv/[slug].ts:18]: This suggests to me that either the filename doesn't matter, or the routing (nearly) doesn't matter.

**Comment by danyx23** [comment]
* **#3603** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3603?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3613** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3613?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danyx23 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by marcelgerber** [comment]
There are some UTF-8 issues with the generated README.
Looking into those, nothing to do there for you for now.

File: functions/_common/grapherRenderer.ts
Comment (marcelgerber) [inline functions/_common/grapherRenderer.ts:1]: Another note for future-us:
We should extract some of this functionality out of `grapherRenderer`, and split it up into some separate files for
- parsing (image) options
- initializing grapher
- generating svg/png
- readme/metadata file generation
- csv file generation

**Review by marcelgerber — APPROVED** [review]
Nice!

I proposed some enhancements in https://github.com/owid/owid-grapher/pull/4051, so please have a look at that one.
Otherwise, this looks really good!

One thing I'm wondering is if in the metadata and README files, we want to note down the date of the download? I think that would be a nice idea.

**Comment by danyx23** [comment]
### Merge activity

* **Oct 10, 12:53 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3613).

**Pull request merged**